asllib/Typing.ml

```diff
@@ -2210,26 +2193,84 @@ module Annotate (C : ANNOTATE_CONFIG) = struct
   (*                                                                            *)
   (******************************************************************************)
 
-  (* Begin Specification *)
-  let type_check_ast ast env =
-    let env = build_global ast env in
+  let type_check_decl d (acc, env) =
+    let here = ASTUtils.add_pos_from_st d in
     let () =
       if false then
         Format.eprintf "@[<v>Typing with %s in env:@ %a@]@." strictness_string
           StaticEnv.pp_env env
     in
-    let annotate d =
-      let here = ASTUtils.add_pos_from_st d in
+    match d.desc with
+    | D_Func ({ body = SB_ASL _; _ } as f) ->
+        let env = declare_one_func d f env in
+        let d = D_Func (try_annotate_subprogram d env f) |> here in
+        (d :: acc, env)
+    | D_Func ({ body = SB_Primitive; _ } as f) ->
+        let env = declare_one_func d f env in
+        let d = D_Func (rename_primitive d env f) |> here in
+        (d :: acc, env)
+    | D_GlobalStorage gsd ->
+        let d = D_GlobalStorage (try_annotate_gsd env gsd) |> here in
+        let env = declare_global_storage d gsd env in
+        (d :: acc, env)
+    | D_TypeDecl (x, ty, s) ->
+        let env = declare_type d x ty s env in
+        (d :: acc, env)
+
+  let type_check_mutually_rec ds (acc, env) =
+    let env =
+      List.fold_left
+        (fun env d ->
+          match d.desc with
+          | D_Func f -> declare_one_func d f env
+          | _ ->
+              fatal_from d
+                (Error.BadRecursiveDecls
+                   (List.map ASTUtils.identifier_of_decl ds)))
+        env ds
+    in
+    let ds =
+      List.map
+        (fun d ->
+          match d.desc with
+          | D_Func ({ body = SB_ASL _; name; _ } as f) ->
+              let () =
+                if false then
+                  Format.eprintf "@[Analysing decl %s.@]@." name
+              in
+              D_Func (try_annotate_subprogram d env f) |> add_pos_from d
+          | D_Func ({ body = SB_Primitive; _ } as f) ->
+              D_Func (rename_primitive d env f) |> add_pos_from d
+          | _ -> assert false)
+        ds
+    in
+    (List.rev_append ds acc, env)
+
+  (* Begin Specification *)
+  let type_check_ast =
+    let def d =
```

Comment:
Again the local function `def` looks the same as `ASTUtils. identifier_of_decl`.